# Проект "Рынок труда"

## Цели и задачи проекта

Цель проекта - проанализировать базу данных вакансий с hh.ru, чтобы понять текущие тенденции на рынке труда для аналитиков.

**Задачи поставленные в проекте:**
- Определить диапазон заработных плат в общем, а именно средние значения, минимумы и максимумы нижних и верхних порогов зарплаты.
- Выявить регионы и компании, в которых сосредоточено наибольшее количество вакансий.
- Проанализировать, какие преобладают типы занятости, а также графики работы.
- Изучить распределение грейдов (Junior, Middle, Senior) среди аналитиков данных и системных аналитиков.
- Определить наиболее востребованные навыки (как жёсткие, так и мягкие) для различных грейдов и позиций.

## Описание данных

База данных, с которой я буду работать, состоит из одной таблицы. В ней собраны объявления, опубликованные на сайте по поиску работы hh.ru с февраля по июнь 2024 года. Таблица хранит данные о работодателе, вакансии, заработной плате, а также основные жёсткие и мягкие навыки, разделённые на разные столбцы.
- `id` — идентификатор объявления (первичный ключ).
- `name` — название вакансии.
- `published_at` — дата выставления вакансии.
- `employer` — компания, которая ищет работника.
- `department` — отделение компании, в котором необходим сотрудник.
- `area` — расположение компании.
- `experience` — необходимый опыт и ранг кандидата.
- `schedule` — график работы.
- `employment` — занятость: полная или частичная.
- `salary_from` — заработная плата в категории «от».
- `salary_to` — заработная плата в категории «до».
- `salary_bin` — заработная плата в категории «зарплатная вилка».
- `key_skills_1` — первый ключевой навык, который требуется от кандидата.
- `key_skills_2` — второй ключевой навык, который требуется от кандидата.
- `key_skills_3` — третий ключевой навык, который требуется от кандидата.
- `key_skills_4` — четвёртый ключевой навык, который требуется от кандидата.
- `soft_skills_1` — первый дополнительный навык, который требуется от кандидата.
- `soft_skills_2` — второй дополнительный навык, который требуется от кандидата.
- `soft_skills_3` — третий дополнительный навык, который требуется от кандидата.
- `soft_skills_4` — четвёртый дополнительный навык, который требуется от кандидата.

## Инструменты анализа

Для анализа данных был использован PostgreSQL со следующими функциями и техниками:
- Агрегатные функции — COUNT, AVG, MIN, MAX для расчёта количества вакансий и статистик по зарплатам
- Функции округления — ROUND для приведения средних значений к читаемому виду
- Оконные функции внутри агрегации — PERCENTILE_CONT(0.5) WITHIN GROUP для расчёта медианы и MODE() WITHIN GROUP для определения наиболее частого значения в разрезе групп
- Группировка данных — GROUP BY (в том числе по нескольким столбцам одновременно) для агрегации
- Сортировка и ограничение выборки — ORDER BY и LIMIT для получения топ-N значений
- Работа с датами — DATE_TRUNC для группировки вакансий по месяцам публикации
- Фильтрация данных — WHERE и оператор LIKE с шаблонами для поиска вакансий по ключевым словам в названии
- Подзапросы и объединение выборок — вложенные SELECT с UNION ALL для разворачивания столбцов в единый список перед подсчётом

## Работа над проектом

### Настройка работы с SQL

In [1]:
%load_ext sql

In [2]:
!pip install psycopg2-binary

In [3]:
import psycopg2
print(psycopg2.__version__)

2.9.12 (dt dec pq3 ext lo64)


### Выгрузка данных

In [4]:
import urllib.parse

user = "praktikum_student"
password = "Sdf4$2;d-d30pp"
host = "rc1b-wcoijxj3yxfsf3fs.mdb.yandexcloud.net"
port = 6432
dbname = "data-analyst-hhru"

encoded_password = urllib.parse.quote_plus(password)
conn_str = f"postgresql+psycopg2://{user}:{encoded_password}@{host}:{port}/{dbname}?sslmode=require"

%sql $conn_str

In [5]:
!pip install "prettytable<3.0.0"

### Знакомство с данными

In [6]:
%%sql
-- Получим метаданные таблицы, чтобы изучить типы данных и ограничения

SELECT column_name, data_type, is_nullable, column_default
FROM information_schema.columns
WHERE table_name = 'parcing_table';

 * postgresql+psycopg2://praktikum_student:***@rc1b-wcoijxj3yxfsf3fs.mdb.yandexcloud.net:6432/data-analyst-hhru?sslmode=require
20 rows affected.


column_name,data_type,is_nullable,column_default
id,integer,NO,None
published_at,timestamp without time zone,YES,None
salary_from,numeric,YES,None
salary_to,numeric,YES,None
department,character varying,YES,None
area,character varying,YES,None
experience,character varying,YES,None
schedule,character varying,YES,None
employment,character varying,YES,None
soft_skills_3,character varying,YES,None


In [7]:
%%sql
-- Общее количество записей

SELECT
  COUNT(*) AS total_rows
FROM public.parcing_table;

 * postgresql+psycopg2://praktikum_student:***@rc1b-wcoijxj3yxfsf3fs.mdb.yandexcloud.net:6432/data-analyst-hhru?sslmode=require
1 rows affected.


total_rows
1801


In [8]:
%%sql
-- Распределение по месяцам

SELECT
  DATE_TRUNC('month', published_at) AS month,
  COUNT(*) AS cnt
FROM public.parcing_table
GROUP BY 1
ORDER BY 1;

 * postgresql+psycopg2://praktikum_student:***@rc1b-wcoijxj3yxfsf3fs.mdb.yandexcloud.net:6432/data-analyst-hhru?sslmode=require
5 rows affected.


month,cnt
2024-02-01 00:00:00,1
2024-03-01 00:00:00,287
2024-04-01 00:00:00,490
2024-05-01 00:00:00,633
2024-06-01 00:00:00,390


In [9]:
%%sql
-- Доля непустых значений по основным полям

SELECT
  ROUND(100.0 * COUNT(salary_from) / COUNT(*), 1) AS pct_salary_from,
  ROUND(100.0 * COUNT(salary_to) / COUNT(*), 1) AS pct_salary_to,
  ROUND(100.0 * COUNT(experience) / COUNT(*), 1) AS pct_experience,
  ROUND(100.0 * COUNT(area) / COUNT(*), 1) AS pct_area,
  ROUND(100.0 * COUNT(employer) / COUNT(*), 1) AS pct_employer
FROM public.parcing_table;

 * postgresql+psycopg2://praktikum_student:***@rc1b-wcoijxj3yxfsf3fs.mdb.yandexcloud.net:6432/data-analyst-hhru?sslmode=require
1 rows affected.


pct_salary_from,pct_salary_to,pct_experience,pct_area,pct_employer
19.3,12.7,100.0,100.0,100.0


Выводы:
- Общий объём выборки - 1801 вакансия
- Пик публикаций приходится на май (633 вакансии), есть, вероятно случайный, выброс в феврале
- Только ~19% вакансий содержат `salary_from`, а `salary_to` - еще меньше (12.7%). Это значит, что любые расчёты средних или медианных зарплат будут основаны на небольшой подвыборке и могут не отражать реальную картину по рынку. Вероятно работодатели часто указывают зарплату по договорённости и не публикуют ее в открытых данных.
- Поля `experience`, `area` и `employer` не содержат пропусков

### Анализ данных

In [10]:
%%sql
-- Диапазон заработных плат:
    
SELECT
    ROUND(AVG(salary_from), 0) AS avg_salary_from,
    ROUND(AVG(salary_to), 0) AS avg_salary_to,
    MIN(salary_from) AS min_salary_from,
    MIN(salary_to) AS min_salary_to,
    MAX(salary_from) AS max_salary_from,
    MAX(salary_to) AS max_salary_to
FROM public.parcing_table
WHERE salary_from > 50
AND name LIKE '%Аналитик%' OR name LIKE '%аналитик%';

 * postgresql+psycopg2://praktikum_student:***@rc1b-wcoijxj3yxfsf3fs.mdb.yandexcloud.net:6432/data-analyst-hhru?sslmode=require
1 rows affected.


avg_salary_from,avg_salary_to,min_salary_from,min_salary_to,max_salary_from,max_salary_to
103769,144391,25000.0,26000.0,350000.0,467500.0


Выводы:
- Средняя зарплата в категории «от» составляет около 103769 рублей, а в категории «до» — около 144391 рублей.
- Это может указывать на то, что работодатели готовы платить аналитикам данных и системным аналитикам в среднем около 120000 рублей, но надо учитывать, что объем данных здесь ограничен.
- Минимальная предлагаемая зарплата начинается с 25000 рублей, а максимальная достигает 467500 рублей.

In [11]:
%%sql
-- Кол-во вакансий по регионам:
SELECT
    area,
    COUNT(*) AS total_vacancies
FROM public.parcing_table
WHERE name LIKE '%Аналитик%' OR name LIKE '%аналитик%'
GROUP BY area
ORDER BY total_vacancies DESC
LIMIT 10;

 * postgresql+psycopg2://praktikum_student:***@rc1b-wcoijxj3yxfsf3fs.mdb.yandexcloud.net:6432/data-analyst-hhru?sslmode=require
10 rows affected.


area,total_vacancies
Москва,941
Санкт-Петербург,136
Екатеринбург,48
Нижний Новгород,33
Владивосток,29
Казань,28
Новосибирск,27
Краснодар,19
Челябинск,10
Пермь,9


Выводы:
- Москва и Санкт-Петербург — лидеры по количеству вакансий. Это неудивительно, учитывая, что это крупнейшие города с развитой инфраструктурой и большим количеством компаний. 
- В Екатеринбурге, Нижнем Новгороде, Владивосток, Казань и Новосибирск также значительное количество вакансий — это указывает на развитый рынок труда для аналитиков данных в этих регионах.

In [12]:
%%sql
-- Кол-во вакансий по работодателям:
SELECT
    employer,
    COUNT(*) AS total_vacancies
FROM public.parcing_table
WHERE name LIKE '%Аналитик%' OR name LIKE '%аналитик%'
GROUP BY employer
ORDER BY total_vacancies DESC
LIMIT 10;

 * postgresql+psycopg2://praktikum_student:***@rc1b-wcoijxj3yxfsf3fs.mdb.yandexcloud.net:6432/data-analyst-hhru?sslmode=require
10 rows affected.


employer,total_vacancies
СБЕР,135
Ozon,34
Банк ВТБ (ПАО),28
Т1,24
"МАГНИТ, Розничная сеть",17
WILDBERRIES,17
МТС,16
Центральный банк Российской Федерации,15
Правительство Москвы,15
Ростелеком,15


Выводы:
- СБЕР предлагает 135 вакансий и является крупнейшим работодателем для аналитиков данных и системных аналитиков. Это может говорить о значительных инвестициях в аналитику и технологические решения в компании. К тому же в банковской среде всегда нужны аналитики.
- Ozon, ВТБ и другие крупные компании также активно ищут специалистов в области данных, что говорит о высоком спросе на аналитиков в крупных корпорациях.

In [13]:
%%sql
-- Кол-во вакансий по графику работы:
SELECT
    schedule,
    COUNT(*) AS total
FROM public.parcing_table
WHERE name LIKE '%Аналитик%' OR name LIKE '%аналитик%'
GROUP BY schedule
ORDER BY total DESC;

 * postgresql+psycopg2://praktikum_student:***@rc1b-wcoijxj3yxfsf3fs.mdb.yandexcloud.net:6432/data-analyst-hhru?sslmode=require
4 rows affected.


schedule,total
Полный день,1157
Удаленная работа,229
Гибкий график,32
Сменный график,8


Выводы:
- Большинство вакансий (1157) предлагают работу с полным днём.
- Однако значительное количество вакансий (229) также позволяет удалённую работу. Это указывает на то, что работодатели готовы быть гибкими в современных условиях.

In [14]:
%%sql
-- Кол-во вакансий по типу занятости:
SELECT
    employment,
    COUNT(*) AS total
FROM public.parcing_table
WHERE name LIKE '%Аналитик%' OR name LIKE '%аналитик%'
GROUP BY employment
ORDER BY total DESC;

 * postgresql+psycopg2://praktikum_student:***@rc1b-wcoijxj3yxfsf3fs.mdb.yandexcloud.net:6432/data-analyst-hhru?sslmode=require
4 rows affected.


employment,total
Полная занятость,1395
Стажировка,15
Частичная занятость,12
Проектная работа,4


Выводы:
- Подавляющее большинство вакансий (1395) предлагают полную занятость. Это указывает на то, что работодатели предпочитают нанимать аналитиков данных и системных аналитиков на постоянные позиции.
- Возможно, дело в необходимости глубоко погружаться в проекты и долго в них участвовать.

In [15]:
%%sql
-- Распределение грейдов и средние данные по ним:
SELECT
    experience,
    COUNT(*) AS total_vacancies,
    ROUND(AVG(salary_from), 0) AS avg_salary_from,
    PERCENTILE_CONT(0.5) WITHIN GROUP (ORDER BY salary_from) AS median_salary_from,
    ROUND(AVG(salary_to), 0) AS avg_salary_to,
    PERCENTILE_CONT(0.5) WITHIN GROUP (ORDER BY salary_to) AS median_salary_to,
    MODE() WITHIN GROUP(ORDER BY schedule) AS most_common_schedule,
    MODE() WITHIN GROUP(ORDER BY employment) AS employment
FROM public.parcing_table
WHERE name LIKE '%Аналитик%' OR name LIKE '%аналитик%'
GROUP BY experience
ORDER BY experience;

 * postgresql+psycopg2://praktikum_student:***@rc1b-wcoijxj3yxfsf3fs.mdb.yandexcloud.net:6432/data-analyst-hhru?sslmode=require
4 rows affected.


experience,total_vacancies,avg_salary_from,median_salary_from,avg_salary_to,median_salary_to,most_common_schedule,employment
Junior (no experince),123,63377,60000.0,73637,80000.0,Полный день,Полная занятость
Junior+ (1-3 years),906,93895,80000.0,124302,110000.0,Полный день,Полная занятость
Middle (3-6 years),391,163344,150000.0,229801,200000.0,Полный день,Полная занятость
Senior (6+ years),6,156667,200000.0,240000,240000.0,Полный день,Полная занятость


Выводы:
- Наибольшее количество вакансий предназначено для специалистов с опытом от 1 до 3 лет (Junior+), что свидетельствует о высоком спросе на специалистов уже продвинутого уровня. 
- Вакансий для Middle-специалистов (3–6 лет) также много, в то время как спрос на Senior-специалистов (6+ лет) крайне низкий, возможно, из-за узкого круга специалистов с таким уровнем опыта или предпочитаемых долгосрочных позиций. Однако важно понимать, что Senior-аналитиков могут чаще искать по другим каналам.
- Также достаточно мало вакансий для специалистов без опыта (Junior), что говорит о сдержаном спросе на таких претендентов. Возможно есть иные каналы для их привлечения, в том числе через стажировки.
- Заметен четкий рост зарплаты исходя из опыта специалиста: Junior+ в среднем имеет зарплату на 25% выше, чем Junior, а Middle почти в 2 раза выше, чем Junior+.
- При этом, исходя из медианных значений, средние арифметические значения несколько завышены небольшим кол-вом выбросов вакансий с зарплатной заметно выше рынка. Вплоть до того, что может создаться впечатление, будто Middle-специалист зарабатывает больше Senior. По медианным значениям это, очевидно, не так.

In [16]:
%%sql
-- Основные работодатели и средние данные по ним:
SELECT
    employer,
    COUNT(*) AS total_vacancies,
    ROUND(AVG(salary_from), 0) AS avg_salary_from,
    ROUND(AVG(salary_to), 0) AS avg_salary_to,
    employment,
    schedule
FROM public.parcing_table
WHERE name LIKE '%Аналитик%' OR name LIKE '%аналитик%'
GROUP BY employer, employment, schedule
ORDER BY total_vacancies DESC
LIMIT 20;

 * postgresql+psycopg2://praktikum_student:***@rc1b-wcoijxj3yxfsf3fs.mdb.yandexcloud.net:6432/data-analyst-hhru?sslmode=require
20 rows affected.


employer,total_vacancies,avg_salary_from,avg_salary_to,employment,schedule
СБЕР,131,110583,73333,Полная занятость,Полный день
Банк ВТБ (ПАО),28,None,None,Полная занятость,Полный день
Ozon,24,None,None,Полная занятость,Полный день
Т1,18,None,None,Полная занятость,Полный день
МТС,16,67500,None,Полная занятость,Полный день
Правительство Москвы,15,None,None,Полная занятость,Полный день
Центральный банк Российской Федерации,15,None,None,Полная занятость,Полный день
Яндекс,15,None,None,Полная занятость,Полный день
Райффайзен Банк,13,None,None,Полная занятость,Полный день
WILDBERRIES,12,None,None,Полная занятость,Полный день


Выводы:
- СБЕР выделяется как основной работодатель для аналитиков данных и системных аналитиков с 131 вакансиями и средней зарплатой от 110583 рублей до 73333 рублей. 
- Средняя зарплата «до» ниже средней зарплаты «от» — скорее всего, СБЕР часто пишет сумму «до», не указывая сумму «от». 
- Основной тип занятости в СБЕР — полная занятость с полным рабочим днём. Значит, компания предпочитает длительное сотрудничество с сотрудниками.
- Банк ВТБ (ПАО) и Ozon также предлагают достаточно вакансий с аналогичными условиями занятости. Это подтверждает тенденцию крупных компаний нанимать аналитиков данных на полную занятость с фиксированным рабочим графиком.
- Если вам подходят такие условия, то потенциальных работодателей для вас будет много.

In [17]:
%%sql
-- Навыки для Junior:
SELECT
  key_skills,
  COUNT(*) AS demand_count
FROM (
  SELECT key_skills_1 AS key_skills FROM public.parcing_table WHERE experience = 'Junior (no experince)' AND name LIKE '%Аналитик%' OR name LIKE '%аналитик%'
  UNION ALL
  SELECT key_skills_2 AS key_skills FROM public.parcing_table WHERE experience = 'Junior (no experince)' AND name LIKE '%Аналитик%' OR name LIKE '%аналитик%'
  UNION ALL
  SELECT key_skills_3 AS key_skills FROM public.parcing_table WHERE experience = 'Junior (no experince)' AND name LIKE '%Аналитик%' OR name LIKE '%аналитик%'
  UNION ALL
  SELECT key_skills_4 AS key_skills FROM public.parcing_table WHERE experience = 'Junior (no experince)' AND name LIKE '%Аналитик%' OR name LIKE '%аналитик%'
) AS t
WHERE key_skills IS NOT NULL AND key_skills != ''
GROUP BY key_skills
ORDER BY demand_count DESC
LIMIT 10;

 * postgresql+psycopg2://praktikum_student:***@rc1b-wcoijxj3yxfsf3fs.mdb.yandexcloud.net:6432/data-analyst-hhru?sslmode=require
10 rows affected.


key_skills,demand_count
SQL,171
Анализ данных,96
Python,96
Документация,68
Коммуникация,43
Power BI,43
MS SQL,37
Pandas,32
Аналитическое мышление,30
MS Excel,22


Выводы:
- Основные хард-навыки для Junior специалиста это SQL, Анализ данных и Python.
- Исходя из этого, работодатели ожидают увидеть начинающих специалистов, которые уже умеют добывать данные из базы данных, очищать их и превращать в полезные инсайты, а не просто работать с уже готовой сводкой.

In [18]:
%%sql
-- Софт-навыки для аналитика:
SELECT
  soft_skills,
  COUNT(*) AS demand_count
FROM (
  SELECT soft_skills_1 AS soft_skills FROM public.parcing_table WHERE name LIKE '%Аналитик%' OR name LIKE '%аналитик%'
  UNION ALL
  SELECT soft_skills_2 AS soft_skills FROM public.parcing_table WHERE name LIKE '%Аналитик%' OR name LIKE '%аналитик%'
  UNION ALL
  SELECT soft_skills_3 AS soft_skills FROM public.parcing_table WHERE name LIKE '%Аналитик%' OR name LIKE '%аналитик%'
  UNION ALL
  SELECT soft_skills_4 AS soft_skills FROM public.parcing_table WHERE name LIKE '%Аналитик%' OR name LIKE '%аналитик%'
) AS t
WHERE soft_skills IS NOT NULL AND soft_skills != ''
GROUP BY soft_skills
ORDER BY demand_count DESC
LIMIT 10;

 * postgresql+psycopg2://praktikum_student:***@rc1b-wcoijxj3yxfsf3fs.mdb.yandexcloud.net:6432/data-analyst-hhru?sslmode=require
10 rows affected.


soft_skills,demand_count
Документация,196
Коммуникация,144
Аналитическое мышление,97
Аналитическое мышление,35
Документация,32
Проактивность,31
Проактивность,14
Креативность,9
Переговоры,6
Адаптивность,4


Выводы:
- Наиболее часто упоминаемые софт-навыки для аналитика это документация, коммуникация и аналитическое мышление.
- Здесь прослеживается желание работодателей найти универсального специалиста, который сможет легко встроится в рабочий процесс, работать в условиях неопределенности и который послужит мостом между данными и бизнесом.

## Итоговые выводы

- **Объeм и качество данных.** Выборка включает 1801 вакансию за март–июнь 2024 года, с пиком публикаций в мае. Поля `salary_from` (19%) и `salary_to` (13%) заполнены слабо, поэтому зарплатные оценки стоит воспринимать как ориентировочные, основанные на неполной подвыборке.
- **Зарплаты.** Средняя предлагаемая зарплата аналитика — примерно 100–145 тыс. рублей (от/до), с разбросом от 25 000 до 467 500 рублей. Зарплата ощутимо растeт с опытом: Junior+ зарабатывает в среднем на 25% больше Junior, а Middle — почти вдвое больше Junior+. При этом медианные значения ниже средних, что говорит о влиянии отдельных высокооплачиваемых выбросов на общую картину.
- **География и работодатели.** Основной спрос сосредоточен в Москве и Санкт-Петербурге, за ними следуют крупные региональные центры (Екатеринбург, Нижний Новгород, Владивосток, Казань, Новосибирск). Крупнейший работодатель — СБЕР, также заметны Ozon и ВТБ, что указывает на высокий спрос на аналитиков в банковском и e-commerce секторах.
- **Условия работы.** Подавляющее большинство вакансий предполагают полную занятость и полный рабочий день, но заметная доля (около 13%) допускает удалeнный формат — работодатели постепенно становятся более гибкими.
- **Грейды.** Наибольший спрос — на специалистов уровня Junior+ (1–3 года опыта) и Middle, тогда как вакансий для Junior без опыта и для Senior заметно меньше. Это может говорить о том, что начинающих и senior-специалистов чаще ищут через другие каналы (стажировки, реферальные программы).
- **Навыки.** Ключевые хард-скиллы — SQL, анализ данных и Python, что подтверждает ожидание от кандидатов навыков самостоятельной работы с базами данных, а не только с готовыми отчeтами. Среди софт-скиллов лидируют документация, коммуникация и аналитическое мышление — работодатели ищут универсальных специалистов, способных быть связующим звеном между данными и бизнесом.
- **Общий вывод.** Рынок труда для аналитиков в 2024 году активно растeт, сосредоточен в крупных городах и крупных компаниях, требует уверенного владения SQL и Python в сочетании с развитыми коммуникативными навыками, а карьерный рост сопровождается заметным увеличением зарплаты.